File: python/279-perfect-squares.py
import math
class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [2**32]*(n+1)
        for i in range(1, n+1):
            if math.sqrt(i)-int(math.sqrt(i)) == 0:
                dp[i] = 1
            else:
                j = 1
                while j*j <= i:
                    dp[i] = min(dp[i], dp[i-j*j] + 1)
                    j += 1
        print(dp)
        return dp[n]
if __name__ == '__main__':
    Solution().numSquares(10)

File: cpp/279-perfect-squares.cpp
class Solution {
public:
    int numSquares(int n) {
        int dp[n+1];
        dp[0] = 0;
        for(int i = 1; i <= n; i++){
            dp[i] = 2 << 15; //init
            for(int j = 1; j*j <= i; j++){
                dp[i] = min(dp[i], dp[i-j*j]+1);
            }
        }
        return dp[n];
    }
};